File: 0877.石子游戏/solution.cpp
const int MAXN=500;
int dp[MAXN][MAXN];
class Solution {
public:
    bool stoneGame(vector<int>& piles) {
        memset(dp,0xc0,sizeof(dp));
        int sz=piles.size();
        for (int i=0;i<sz;i++) {
            dp[i][i]=piles.at(i);
        }
        return dfs(piles,0,sz-1)>0;
    }
    int dfs(vector<int> &piles,int st,int ed) {
        if (dp[st][ed]!=0xc0c0c0c0) {
            return dp[st][ed];
        }
        dp[st][ed]=max(
            piles.at(st)-dfs(piles,st+1,ed),
            piles.at(ed)-dfs(piles,st,ed-1)
        );
        return dp[st][ed];
    }
};

File: 0877.石子游戏/solution.go
func stoneGame(piles []int) bool {
	sz := len(piles)
	dp := make([][]int, sz)
	for i := range dp {
		dp[i] = make([]int, sz)
		for j := range dp[i] {
			dp[i][j] = math.MinInt32 / 2
		}
		dp[i][i] = piles[i]
	}
	return dfs(dp, piles, 0, sz-1) > 0
}

func dfs(dp [][]int, piles []int, st, ed int) int {
	if dp[st][ed] != math.MinInt32/2 {
		return dp[st][ed]
	}
	dp[st][ed] = max(
		piles[st]-dfs(dp, piles, st+1, ed),
		piles[ed]-dfs(dp, piles, st, ed-1),
	)
	return dp[st][ed]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}